Partial points scoring

MattyMay opened the pull request:
Add partial credit scoring capabilities

MattyMay committed:
add fields to AgTestCommand for partial credit scoring configuration

MattyMay committed:
Implement and test partial credit scoring

MattyMay committed:
Truncate to max partial credit points in submission feedback instead of
in test command result

MattyMay committed:
Fix new test regressions

MattyMay committed:
Don't grade partial credit when test command times out

james-perretta commented on autograder/core/submission_feedback.py:
```diff
@@ -1190,20 +1210,64 @@ def stderr_points_possible(self) -> int:
+                > self._cmd.max_points_for_partial_credit):
+            return self._cmd.max_points_for_partial_credit
+
+        return self._ag_test_command_result.partial_credit_points
```
Could this use `min` instead?

james-perretta commented on autograder/core/submission_feedback.py:
```diff
@@ -1190,20 +1210,64 @@ def stderr_points_possible(self) -> int:
+
+        match self._ag_test_command_result.partial_credit_error:
+            case PartialCreditError.none:
+                return 'none'
```
Maybe this should just be an empty string? I could see this getting interpreted as "there's an error that has to do with None"

james-perretta commented on autograder/core/tests/test_models/test_ag_test/test_ag_test_command.py:
```diff
@@ -798,3 +807,36 @@ def test_stdin_stdout_stderr_source_only_set_to_none_if_source_is_instructor_fil
                          self.ag_test_command.expected_stderr_source)
         self.assertIsNone(self.ag_test_command.expected_stderr_instructor_file)
+
+    def test_partial_credit_source_stdout_when_expected_stdout_source_is_not_none(self):
```
I usually include a check that the appropriate field is in the validation error dictionary. Search the codebase for `as cm:` to see an example

james-perretta commented on autograder/core/tests/test_models/test_ag_test/test_ag_test_command.py:
```diff
@@ -798,3 +807,36 @@ def test_stdin_stdout_stderr_source_only_set_to_none_if_source_is_instructor_fil
+                partial_credit_source=ag_models.PartialCreditSource.stderr
+            )
+
+    def test_partial_credit_source_stdout_when_expected_stderr_source_is_not_none(self):
```
Can we add a test for the fourth combination? (partial credit source stderr, expected stdout not none)

james-perretta commented on autograder/core/tests/test_submission_feedback/test_ag_test_command_result_feedback.py:
```diff
@@ -140,6 +191,136 @@ def test_points_everything_incorrect_max_fdbk(self):
+        self.ag_test_command = obj_build.make_stdout_partial_credit_test_command()
+        cmd_result = self.make_incorrect_result()
+
+        cmd_result.partial_credit_points = - 42
```
Formatting: remove space between - and 42

james-perretta commented on autograder/grading_tasks/tasks/grade_ag_test.py:
```diff
@@ -258,6 +259,34 @@ def grade_ag_test_command_impl(sandbox: AutograderSandbox,
+            else:
+                partial_credit_source = run_result.stderr
+
+            with open(partial_credit_source.name) as f:
```
Note to self: discuss caveats

MattyMay committed:
Fix tests where max_points_for_partial_credit was 0

Fix test_points_everything_correct partial credit tests

Fix tests where max_points_for_partial_credit was 0

james-perretta commented on autograder/core/tests/test_submission_feedback/test_ag_test_command_result_feedback.py:
```diff
@@ -140,6 +195,148 @@ def test_points_everything_incorrect_max_fdbk(self):
+            self.ag_test_case,
+            max_points_for_partial_credit=2
+        )
+        cmd_result = self.make_incorrect_result()
```
Optional: make this function (and the underlying make_incorrect_ag_test_command_result that it calls) take in an argument to specify partial credit points (default to 0)

james-perretta commented on autograder/core/tests/test_submission_feedback/test_ag_test_command_result_feedback.py:
```diff
@@ -949,6 +1146,54 @@ def test_student_on_fail_description_shown_on_fail(self) -> None:
+    def test_student_on_fail_description_shown_on_fail_with_partial_credit(self):
+        description = 'description WAAAAAA'
+        on_fail_description = 'WAAAA fail'
+        normal_fdbk_config = self.ag_test_command.normal_fdbk_config
```
I think this could be passed to the call to `make_stdout_partial_credit_test_command` as a kwarg instead

james-perretta commented on autograder/core/tests/test_submission_feedback/test_ag_test_command_result_feedback.py:
```diff
@@ -949,6 +1146,54 @@ def test_student_on_fail_description_shown_on_fail(self) -> None:
+        self.ag_test_command.validate_and_update(normal_fdbk_config=normal_fdbk_config)
+        for (
+            return_code_correct, stderr_correct, partial_credit_correct
+        ) in itertools.product(opts, opts, opts):
```
I think it's fine to inline the argument to `itertools.product` in both places--makes it clearer what their purpose is. `itertools.product(*itertools.repeat([True, False, None], 3))` should do the trick, but I'm open to alternatives

james-perretta commented on autograder/grading_tasks/tests/test_tasks/test_grade_ag_test.py:
```diff
@@ -237,6 +237,94 @@ def test_wrong_expected_stderr_instructor_file(self, *args):
+        tasks.grade_submission_task(self.submission.pk)
+
+        res = ag_models.AGTestCommandResult.objects.get(ag_test_command=cmd)
+        self.assertEqual(res.partial_credit_points, 2)
```
Reminder: add tests to ag test command feedback for taking the max if the score reported goes over

james-perretta commented on autograder/grading_tasks/tests/test_tasks/test_grade_ag_test.py:
```diff
@@ -400,6 +488,28 @@ def test_program_times_out(self, *args):
+        cmd = obj_build.make_stdout_partial_credit_test_command(
+            self.ag_test_case,
+            cmd=self.timeout_cmd,
+            time_limit=1)
```
Set max partial credit points to nonzero

james-perretta commented on autograder/grading_tasks/tests/test_tasks/test_grade_ag_test.py:
```diff
@@ -400,6 +488,28 @@ def test_program_times_out(self, *args):
+
+        res = ag_models.AGTestCommandResult.objects.get(ag_test_command=cmd)
+        self.assertTrue(res.timed_out)
+        self.assertEqual(res.partial_credit_error, ag_models.PartialCreditError.none)
```
Add check that partial credit points are zero

james-perretta commented on autograder/grading_tasks/tests/test_tasks/test_grade_ag_test.py:
```diff
@@ -400,6 +488,28 @@ def test_program_times_out(self, *args):
+        cmd = obj_build.make_stderr_partial_credit_test_command(
+            self.ag_test_case,
+            cmd=self.timeout_cmd,
+            time_limit=1)
```
Same as above

james-perretta commented on autograder/utils/testing/model_obj_builders.py:
```diff
@@ -412,6 +412,63 @@ def make_full_ag_test_command(
 
+def make_stdout_partial_credit_test_command(
+    ag_test_case: ag_models.AGTestCase | None = None,
+    max_points_for_partial_credit: int | None = None,
```
Suggestion: have this default to some non-zero number, since everywhere this gets called passes in a value

james-perretta commented on autograder/core/tests/test_submission_feedback/test_ag_test_command_result_feedback.py:
```diff
@@ -140,6 +195,148 @@ def test_points_everything_incorrect_max_fdbk(self):
+
+        self.assertEqual(self.min_points_possible - 1337, fdbk.total_points)
+        self.assertEqual(self.max_points_possible, fdbk.total_points_possible)
+
```
Add test for points hidden? (check points, max points, and error)

james-perretta commented on autograder/core/submission_feedback.py:
```diff
@@ -1190,20 +1210,64 @@ def stderr_points_possible(self) -> int:
+            case PartialCreditError.non_integer:
+                return 'Non-integer value found for partial credit score'
+            case PartialCreditError.exceeded_max_points:
+                return ('Partial credit points exceeded the number of partial credit'
```
Idea: include the points parsed in this message?

MattyMay committed:
Make requested changes

MattyMay committed:
Fix test regressions

MattyMay committed:
Change partial_credit_points default to 0. Test that it is 0 when
command times out

MattyMay committed:
Parse output line-by-line for partial credit, add test for non-utf-8
chars

MattyMay committed:
Fix non-utf-8 tests and don't use readlines

MattyMay committed:
Fix test regressions
